TypeCobol.LanguageServer/Processor/DataLayoutProcessor.cs

```diff
@@ -155,29 +159,37 @@ static object GetEnd(long start, long length)
             }
         }
 
-        private List<Tuple<int, DataDefinition, int>> CollectDataLayoutNodes(CompilationUnit compilationUnit)
+        private List<Tuple<int, DataDefinition, int>> CollectDataLayoutNodesAtPosition(CompilationUnit compilationUnit, Position position)
         {
+            var location = CodeElementLocator.FindCodeElementAt(compilationUnit, position);
+            if (location.CodeElement == null)
+            {
+                throw new Exception($"No program found in: '{compilationUnit.TextSourceInfo.Name}'"); ;
+            }
+
             var dataLayoutNodes = new List<Tuple<int, DataDefinition, int>>();
-            Node dataDivision = compilationUnit?.TemporaryProgramClassDocumentSnapshot?.Root?.MainProgram?.GetChildren<DataDivision>()?.FirstOrDefault();
+            var program = location.Node.GetProgramNode();
+            Node dataDivision = program?.GetChildren<DataDivision>()?.FirstOrDefault();
             if (dataDivision != null)
             {
-                // Consider data declared in the Working storage
-                var workingStorage = dataDivision.GetChildren<WorkingStorageSection>().FirstOrDefault();
-                if (workingStorage != null)
-                {
-                    CollectDataLayoutNodes(0, workingStorage, 0);
-                }
+                // Consider data declared in the Working and Local storage
+                CollectDataLayoutNodesInSection<WorkingStorageSection>();
+                CollectDataLayoutNodesInSection<LocalStorageSection>();
+                // Consider also data declared in the Linkage
+                CollectDataLayoutNodesInSection<LinkageSection>();
+            }
 
-                // Consider also data declared in the Local storage
-                var localStorage = dataDivision.GetChildren<LocalStorageSection>().FirstOrDefault();
-                if (localStorage != null)
+            return dataLayoutNodes;
+
+            void CollectDataLayoutNodesInSection<N>() where N : Node
+            {
+                var section = dataDivision.GetChildren<N>().FirstOrDefault();
```

Comment:
Use `Children.OfType<N>().SingleOrDefault()`. `GetChildren` does a copy which is no needed here. Also note that we will iterate over children three times. There is method `ExtractDataSections` in POC_SoDisplay that avoid the multiple iterations but we'll have to find a location for it

Reply:
New properties on DataDivision to store each section